framework/include/times/Times.h

```diff
@@ -0,0 +1,76 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+// Moose includes
+#include "GeneralReporter.h"
+
+/**
+ * Times objects are under the hood Reporters, but limited to a vector of Real
+ */
+class Times : public GeneralReporter
+{
+public:
+  static InputParameters validParams();
+  Times(const InputParameters & parameters);
+  virtual ~Times() = default;
+
+  /// Getter for the full times vector
+  const std::vector<Real> & getTimes() const;
+
+  /// Getter for a set of the full times
+  const std::set<Real> getUniqueTimes() const
+  {
+    return std::set<Real>(_times.begin(), _times.end());
+  };
```

Comment:
what is the value in having duplicate entries? I guess my inclination would be that we just have the set

Reply:
keep our options open?
and sets are both harder to work with as a dev, and slower in the code

Reply:
What APIs on a set are confusing? And speed really depends on what you're doing. If you have to sort a vector a lot, then that will be slow. If you need to search for something in a vector vs a set, its O(N) vs. O(log N)

Reply:
you cant index a set directly for example

my experience with sets is that you always end up converting them to vectors somewhere along the way

"uniqueness" in floating point numbers is also not very well defined with the floating point comparisons issues

Reply:
If you have multiple sources though, or even a single source, what is the usefulness of a `getTimeAtIndex` API? Seems like the `getPreviousTime` and `getNextTime` are the useful APIs and those are only useful if your data is sorted

Reply:
> my experience with sets is that you always end up converting them to vectors somewhere along the way

I don't share this experience

Reply:
And those APIs would require container searches which is O(N) for vectors

Reply:
> > my experience with sets is that you always end up converting them to vectors somewhere along the way
> 
> I don't share this experience

Checkout the block restrictable interface. Under the hood it's a set, and there's vector conversions APIs all over.

Reply:
Sure that's an instance of it. But that's because people designed a lot of APIs that expected vectors instead of sets. Poor API design. In the end we store both the moose and libmesh subdomains as sets

Reply:
"Build a `set` then convert to a sorted `vector`" is almost a C++ design pattern, as the fastest way to handle the "do lots of inserts with duplication, then do lots of searches" use case.

But if neither the inserts nor the searches are in an inner kernel then the speed of sorted vectors doesn't matter.

I'm also not seeing what the use is of getting a time by index.  Knowing an index would imply knowing about every other bit of code that might be adding times, right?  And half the point of this system is to make that unnecessary?